exchanges/bittrex/bittrex_types.go

```diff
@@ -1,226 +1,287 @@
+}
+
+// WsSignalRHandshakeData holds data for the SignalR websocket wrapper handshake
+type WsSignalRHandshakeData struct {
+	URL                     string  `json:"Url"`                     // Path to the SignalR endpoint
+	ConnectionToken         string  `json:"ConnectionToken"`         // Connection token assigned by the server
+	ConnectionID            string  `json:"ConnectionId"`            // The ID of the connection
+	KeepAliveTimeout        float64 `json:"KeepAliveTimeout"`        // Representing the amount of time to wait before sending a keep alive packet over an idle connection
+	DisconnectTimeout       float64 `json:"DisconnectTimeout"`       // Represents the amount of time to wait after a connection goes away before raising the disconnect event
+	ConnectionTimeout       float64 `json:"ConnectionTimeout"`       // Represents the amount of time to leave a connection open before timing out
+	TryWebSockets           bool    `json:"TryWebSockets"`           // Whether the server supports websockets
+	ProtocolVersion         string  `json:"ProtocolVersion"`         // The version of the protocol used for communication
+	TransportConnectTimeout float64 `json:"TransportConnectTimeout"` // The maximum amount of time the client should try to connect to the server using a given transport
+	LongPollDelay           float64 `json:"LongPollDelay"`           // The time to tell the browser to wait before reestablishing a long poll connection after data is sent from the server.
+}
+
+// WsEventRequest holds data on websocket requests
+type WsEventRequest struct {
+	Hub          string      `json:"H"`
+	Method       string      `json:"M"`
+	Arguments    interface{} `json:"A"`
+	InvocationID int         `json:"I"`
+}
+
+// WsEventStatus holds data on the websocket event status
+type WsEventStatus struct {
+	Success   bool   `json:"Success"`
+	ErrorCode string `json:"ErrorCode"`
+}
+
+// WsEventResponse holds data on the websocket response
+type WsEventResponse struct {
+	C string      `json:"C"`
+	S int         `json:"S"`
+	G string      `json:"G"`
+	R interface{} `json:"R"`
+	I int         `json:"I,string"`
+	M []struct {
+		H string   `json:"H"`
+		M string   `json:"M"`
+		A []string `json:"A"`
+	} `json:"M"`
+}
+
+// WsSubscriptionResponse holds data on the websocket response
+type WsSubscriptionResponse struct {
+	C string          `json:"C"`
+	S int             `json:"S"`
+	G string          `json:"G"`
+	R []WsEventStatus `json:"R"`
+	I int             `json:"I,string"`
+	M []struct {
+		H string   `json:"H"`
+		M string   `json:"M"`
+		A []string `json:"A"`
+	} `json:"M"`
+}
+
+// WsAuthResponse holds data on the websocket response
+type WsAuthResponse struct {
+	C string        `json:"C"`
+	S int           `json:"S"`
+	G string        `json:"G"`
+	R WsEventStatus `json:"R"`
+	I int           `json:"I,string"`
+	M []struct {
+		H string   `json:"H"`
+		M string   `json:"M"`
+		A []string `json:"A"`
+	} `json:"M"`
+}
```

Comment:
`WsSubscriptionResponse` and `WsAuthResponse` are different: they have `R []WsEventStatus` and `R WsEventStatus` respectively

Sadly the least contrapted way to figure out which of the two event responses has been received, is to unmarshal the data twice (see `wsHandleResponseData()`. The flow is as follows:
- Unmarshal into WsEventResponse
- Use WsEventResponse.I (InvocationID) to see if it is a response to a subscribe/unsubscribe request or to an auth request
- Unmarshal again into the proper struct

Initially I used type casting to get the array of statuses, but I doubt that's faster - and for sure less readable. E.g., in case of the WsSubscriptionResponse:
- Cast the interface{} to []interface{}
- Iterate the range
- For each item, cast one named array value to a bool and another to a string

I'd like to only unmarshal the R field, but I don't see how I can easily get the raw data of the R field instead of the array that underlies the interface{}

(I'll expand the field names)

Reply:
What you have done seems logical for the situation. Thank you for clarification and sorry for not noticing the difference.